guide/scaling-up/ssr.md (#291)
vuejs-translations/docs-fr: src/guide/scaling-up/ssr.md

### Pourquoi le SSR ? {#why-ssr}

Par rapport à une application côté client Single-Page (SPA), l'avantage du SSR réside principalement dans :

- **Temps d'affichage plus rapide** : cela est plus important sur une connexion internet lente ou sur des dispositifs lents. Le balisage rendu côté serveur n'a pas besoin d'attendre que tout le JavaScript soit téléchargé et exécuté pour être affiché, de sorte que votre utilisateur verra une page entièrement rendue plus rapidement. De plus, la récupération de données est effectuée côté serveur lors de la première visite, ce qui a probablement une connexion plus rapide à votre base de données que le client. Cela entraîne généralement une amélioration des métriques [Core Web Vitals](https://web.dev/vitals/), une meilleure expérience utilisateur et peut être critique pour les applications où le temps d'affichage est directement associé au taux de conversion.

- **Modèle mental unifié** : vous pouvez utiliser le même langage et le même modèle mental déclaratif orienté composant pour développer votre application entière, au lieu de sauter sans arrêt entre un système de modèle de backend et un framework frontend.

> - **Meilleur référencement** : les crawlers de moteur de recherche verront directement la page entièrement rendue.

Author: Je me suis posé la question aussi o.o @edimitchel un avis ?

> - **Meilleur référencement** : les crawlers de moteur de recherche verront directement la page entièrement rendue.

Reviewer: plutôt pour la traduction vu la traduction que tu proposes @tlagoda !

})
```

Après, lancez :

```sh
> node example.js
```

> Cela devrait imprimer ce qui suit sur la ligne de commande :

Reviewer: ```suggestion
Cela devrait retourner :
```

})
```

Enfin, exécutez `node server.js` et visitez `http://localhost:3000`. Vous devriez voir la page fonctionner avec le bouton.

[Essayez-le sur StackBlitz](https://stackblitz.com/fork/vue-ssr-example-basic?file=index.js)

### Hydratation des clients {#client-hydration}

Si vous cliquez sur le bouton, vous verrez que le nombre ne change pas. L'HTML est complètement statique côté client car nous n'avons pas chargé Vue dans le navigateur.

> Pour rendre l'application côté client interactive, Vue doit effectuer l'étape de **hydratation**. Pendant l'hydratation, elle crée la même application Vue qui a été exécutée sur le serveur, associe chaque composant aux nœuds DOM qu'il doit contrôler et attache les écouteurs d'événements du DOM.

Reviewer: ```suggestion
Pour rendre l'application côté client interactive, Vue doit effectuer l'étape de **hydratation**. Pendant l'hydratation, il crée la même application Vue qui a été exécutée sur le serveur, associe chaque composant aux nœuds DOM qu'il doit contrôler et attache les écouteurs d'événements du DOM.
```

Pour monter une application en mode hydratation, nous devons utiliser [`createSSRApp()`](/api/application.html#createssrapp) au lieu de `createApp()`:

```js{2}
// cela fonctionne dans le navigateur.
import { createSSRApp } from 'vue'

const app = createSSRApp({
  // ...même application que sur le serveur
})

// Le montage d'une application SSR sur le client suppose
// que le HTML a été pré-rendu et qu'il effectuera une

> // hydratation au lieu de monter de nouveaux noeuds du DOM.

Reviewer: ```suggestion
// hydratation au lieu de monter de nouveaux nœuds du DOM.
```

})
```

Enfin, exécutez `node server.js` et visitez `http://localhost:3000`. Vous devriez voir la page fonctionner avec le bouton.

[Essayez-le sur StackBlitz](https://stackblitz.com/fork/vue-ssr-example-basic?file=index.js)

### Client Hydration {#client-hydration}

Si vous cliquez sur le bouton, vous verrez que le nombre ne change pas. L'HTML est complètement statique côté client car nous n'avons pas chargé Vue dans le navigateur.

> Pour rendre l'application côté client interactive, Vue doit effectuer l'étape de **hydratation**. Pendant l'hydratation, elle crée la même application Vue qui a été exécutée sur le serveur, associe chaque composant aux nœuds DOM qu'il doit contrôler et attache les écouteurs d'événements DOM.

Author: @tlagoda j'ai mis au féminin car on parle de "l'application Vue" non?

> Pour rendre l'application côté client interactive, Vue doit effectuer l'étape de **hydratation**. Pendant l'hydratation, elle crée la même application Vue qui a été exécutée sur le serveur, associe chaque composant aux nœuds DOM qu'il doit contrôler et attache les écouteurs d'événements DOM.

Reviewer 2: Pour moi le sujet c'est Vue nop ?

> Pour rendre l'application côté client interactive, Vue doit effectuer l'étape de **hydratation**. Pendant l'hydratation, elle crée la même application Vue qui a été exécutée sur le serveur, associe chaque composant aux nœuds DOM qu'il doit contrôler et attache les écouteurs d'événements DOM.

Reviewer: Je suis de l'avis de @tlagoda :)

})
```

De plus, pour charger les fichiers client dans le navigateur, nous devons également :

1. Servir les fichiers client en ajoutant `server.use(express.static('.'))` dans `server.js`.
2. Charger l'entrée client en ajoutant `<script type="module" src="/client.js"></script>` au fichier HTML.
3. Prend en charge l'utilisation comme `import * from 'vue'` dans le navigateur en ajoutant [Import Map](https://github.com/WICG/import-maps) au fichier HTML.

[Essayez l'exemple complété sur StackBlitz](https://stackblitz.com/fork/vue-ssr-example?file=index.js). Le bouton est maintenant interactif !

## Solutions de haut niveau {#higher-level-solutions}

Pour une application SSR en production, plusieurs considérations supplémentaires doivent être prises en compte :

- Prendre en charge les SFC Vue et d'autres exigences de l'étape de build. En fait, nous aurons besoin de coordonner deux builds pour la même application : un pour le client et un pour le serveur.

  :::tip
  Les composants Vue sont compilés différemment en SSR : les templates compilés sont compilés en chaînes, par rapport à des fonctions de rendu de DOM virtuel, pour une performance accrue.
  :::

> - Dans le gestionnaire de requête du serveur, afficher le HTML avec les liens de ressources client corrects et les astuces de ressources optimales. Il peut également être nécessaire de basculer entre le mode SSR et SSG, ou même de mélanger les deux dans la même application.

Reviewer: ```suggestion
- Dans le gestionnaire de requête du serveur, afficher le HTML avec les liens corrects des ressources clientes et autres ressources optimisées. Il peut également être nécessaire de basculer entre le mode SSR et SSG, ou même de mélanger les deux dans la même application.
```

- Gestion du routage, de la récupération de données et des stores de gestion d'état de manière universelle.

Une implémentation complète serait assez complexe et dépend de la chaîne d'outils de build que vous avez choisie. Par conséquent, nous vous recommandons fortement d'adopter une solution plus élevée et orientée afin d'abstraire la complexité pour vous. Ci-dessous, nous présenterons quelques solutions SSR recommandées dans l'écosystème Vue.

### Nuxt {#nuxt}

> [Nuxt](https://nuxt.com/) Nuxt est une plateforme Vue simplifiée pour développer des applications universelles et peut être utilisée comme générateur de site statique. Nous vous recommandons fortement de l'essayer.

Reviewer: ```suggestion
[Nuxt](https://nuxt.com/) est une plateforme Vue simplifiée pour développer des applications universelles et peut être utilisée comme générateur de site statique. Nous vous recommandons fortement de l'essayer.
```

### Quasar {#quasar}

[Quasar](https://quasar.dev) est une plateforme basée sur Vue qui vous permet de développer des applications pour plusieurs plateformes (SSR, SPA, PWA, mobile, bureau, extension de navigateur) avec un seul code de base. Il fournit également des composants d'interface utilisateur conformes à Material Design.

### Vite SSR {#vite-ssr}

> Vite fournit [un support intégré pour le rendu côté serveur de Vue](https://vitejs.dev/guide/ssr.html), mais il est intentionnellement de bas niveau. Si vous souhaitez utiliser directement avec Vite, consultez [vite-plugin-ssr](https://vite-plugin-ssr.com/), un plugin communautaire qui fait abstraction de nombreux détails complexes pour vous.

Reviewer: ```suggestion
Vite fournit [un support natif pour le rendu côté serveur de Vue](https://vitejs.dev/guide/ssr.html), mais il est intentionnellement de bas niveau. Si vous souhaitez utiliser directement avec Vite, consultez [vite-plugin-ssr](https://vite-plugin-ssr.com/), un plugin communautaire qui fait abstraction de nombreux détails complexes pour vous.
```

Vous pouvez également trouver un exemple de projet Vue + Vite SSR en utilisant une configuration manuelle ici, qui peut servir de base pour le build. Notez que ceci n'est recommandé que si vous avez de l'expérience avec SSR / outils de build et que vous souhaitez avoir un contrôle complet sur l'architecture de haut niveau.

> ## Écrire un code convivial pour du SSR {#writing-ssr-friendly-code}

Reviewer: ```suggestion
## Écrire du code SSR propre {#writing-ssr-friendly-code}
```
Convivial me semble pas correct ici

Indépendamment de votre configuration de build ou de votre choix de framework de haut niveau, il existe des principes qui s'appliquent à toutes les applications Vue SSR.

### Réactivité sur le serveur {#reactivity-on-the-server}

Pendant le SSR, chaque URL requêtée correspond à un état souhaité de notre application. Il n'y a pas d'interaction utilisateur et aucune mise à jour du DOM, donc la réactivité n'est pas nécessaire sur le serveur. Par défaut, la réactivité est désactivée pendant le SSR pour une meilleure performance.

### Les hooks du cycle de vie du composant {#component-lifecycle-hooks}

> Depuis qu'il n'y a plus de mises à jour dynamiques, les hooks de cycle de vie tels que <span class="options-api">`mounted`</span><span class="composition-api">`onMounted`</span> ou <span class="options-api">`updated`</span><span class="composition-api">`onUpdated`</span> ne seront **PAS** appelés pendant le SSR et ne seront exécutés que côté client.<span class="options-api"> Les seuls hooks qui sont appelés pendant le SSR sont `beforeCreate` et `created`</span>

Reviewer: ```suggestion
Vu qu'il n'y a pas de mises à jour dynamiques, les hooks de cycle de vie tels que <span class="options-api">`mounted`</span><span class="composition-api">`onMounted`</span> ou <span class="options-api">`updated`</span><span class="composition-api">`onUpdated`</span> ne seront **PAS** appelés pendant le SSR et ne seront exécutés que côté client.<span class="options-api"> Les seuls hooks qui sont appelés pendant le SSR sont `beforeCreate` et `created`</span>
```

Notez que si une bibliothèque tiers n'est pas écrite en vue d'une utilisation universelle, il peut être difficile de l'intégrer dans une application rendue côté serveur. Vous _pourriez_ être en mesure de la faire fonctionner en simulant certaines des globales, mais ce serait un bricolage et pourrait interférer avec le code de détection de l'environnement d'autres bibliothèques.

### Pollution d'État à demande croisée {#cross-request-state-pollution}

Dans le chapitre de Gestion d'État, nous avons introduit un [patron de gestion d'état simple en utilisant les API de réactivité](state-management.html#simple-state-management-with-reactivity-api). Dans un contexte SSR, ce modèle nécessite des ajustements supplémentaires.

Le modèle déclare un état partagé dans la portée d'un module JavaScript. Cela les rend **singletons** - c'est-à-dire qu'il n'y a qu'une seule instance de l'objet réactif tout au long de la vie de notre application. Cela fonctionne comme prévu dans une application Vue côté client, car les modules de notre application sont initialisés à nouveau pour chaque visite de page navigateur.

> Cependant, dans un contexte SSR, les modules d'application sont généralement initialisés une seule fois sur le serveur, lorsque le serveur démarre. Les mêmes instances de modules seront réutilisées à travers plusieurs demandes de serveur, et donc nos objets d'état singletons. Si nous modifions l'état partagé singleton avec des données spécifiques à un utilisateur, cela peut être accidentellement fuir vers une demande provenant d'un autre utilisateur. Nous appelons cela une **pollution d'état cross-request.**

Reviewer: ```suggestion
Cependant, dans un contexte SSR, les modules d'application sont généralement initialisés une seule fois sur le serveur, lorsque le serveur démarre. Les mêmes instances de modules seront réutilisées à travers plusieurs demandes de serveur, et donc nos objets d'état singletons. Si nous modifions l'état partagé singleton avec des données spécifiques à un utilisateur, cela peut être accidentellement fuir vers une demande provenant d'un autre utilisateur. Nous appelons cela une **pollution d'état par demandes croisées.**
```

Techniquement, nous pouvons ré-initialiser tous les modules JavaScript à chaque demande, tout comme nous le faisons dans les navigateurs. Cependant, l'initialisation des modules JavaScript peut être coûteuse, ce qui affecterait significativement les performances du serveur.

La solution recommandée est de créer une nouvelle instance de l'application entière - y compris le routeur et les stores globaux - à chaque demande. Ensuite, au lieu de l'importer directement dans nos composants, nous fournissons l'état partagé en utilisant [Provide au niveau de l'application](/guide/components/provide-inject.html#app-level-provide) et l'injecterons dans les composants qui en ont besoin :

```js
// app.js (partagé entre le serveur et le client)
import { createSSRApp } from 'vue'
import { createStore } from './store.js'

> // appelé sur chaque demande

Reviewer: ```suggestion
// appelée sur chaque demande
```

export function createApp() {
  const app = createSSRApp(/* ... */)
  // crée une nouvelle instance de store par requête
  const store = createStore(/* ... */)

>   // fournis un store au niveau de l'application

Reviewer: ```suggestion
  // fournit un store au niveau de l'application
```

})
```

De plus, pour charger les fichiers client dans le navigateur, nous devons également :

1. Servir les fichiers client en ajoutant `server.use(express.static('.'))` dans `server.js`.
2. Charger l'entrée client en ajoutant `<script type="module" src="/client.js"></script>` à la coque HTML.

> 3. Prend en charge l'utilisation comme `import * from 'vue'` dans le navigateur en ajoutant [Import Map](https://github.com/WICG/import-maps) à la coque HTML.

Reviewer 2: ```suggestion
3. Prendre en charge les cas d'usage tels que `import * from 'vue'` dans le navigateur en ajoutant [Import Map](https://github.com/WICG/import-maps) à la coque HTML.
```
Pas sûr pour coque HTML ?

Author: @tlagoda Oui bizarre qu'il l'ait nommé comme ça. Jamais vu. J'aurais dis au fichier ou à l'interface HTML ?

Reviewer 2: fichier me paraît pas mal, voir html tout court

> [Essayez l'exemple complété sur StackBlitz](https://stackblitz.com/fork/vue-ssr-example?file=index.js).  Le bouton est maintenant interactif !

Reviewer 2: ```suggestion
[Essayez l'exemple complété sur StackBlitz](https://stackblitz.com/fork/vue-ssr-example?file=index.js). Le bouton est maintenant interactif !
```

## Solutions de haut niveau {#higher-level-solutions}

Lorsque vous passez d'un exemple à une application SSR en production, il y a beaucoup plus de considérations à prendre en compte. Nous aurons besoin de :

> - Prise en charge des SFC Vue et d'autres exigences de la phase de construction. En fait, nous aurons besoin de coordonner deux builds pour la même application: un pour le client et un pour le serveur.

Reviewer 2: ```suggestion
- Prendre en charge les SFC Vue et d'autres exigences de l'étape de build. En fait, nous aurons besoin de coordonner deux builds pour la même application : un pour le client et un pour le serveur.
```

  :::tip

>   Les composants Vue sont compilés différemment lorsqu'ils sont utilisés en SSR - les modèles sont compilés en concaténations de chaînes au lieu de fonctions de rendu de DOM virtuel pour un rendu plus efficace en performance.

Reviewer 2: ```suggestion
  Les composants Vue sont compilés différemment lorsqu'ils sont utilisés en SSR - les templates sont compilés en concaténations de chaînes au lieu de fonctions de rendu de DOM virtuel pour un rendu plus efficace en performance.
```

  :::

- Dans le gestionnaire de requête du serveur, afficher le HTML avec les liens de ressources client corrects et les astuces de ressources optimales. Il peut également être nécessaire de basculer entre le mode SSR et SSG, ou même de mélanger les deux dans la même application.

> - Gestion du routage, de la récupération de données et des magasins de gestion d'état de manière universelle.

Reviewer 2: ```suggestion
- Gestion du routage, de la récupération de données et des stores de gestion d'état de manière universelle.
```

Une implémentation complète serait assez complexe et dépend de la chaîne d'outils de build que vous avez choisie. Par conséquent, nous vous recommandons fortement d'adopter une solution plus élevée et orientée afin d'abstraire la complexité pour vous. Ci-dessous, nous présenterons quelques solutions SSR recommandées dans l'écosystème Vue.

### Nuxt {#nuxt}

[Nuxt](https://nuxt.com/) est une plateforme de haut niveau construite sur l'écosystème Vue qui offre une expérience de développement simplifiée pour écrire des applications Vue universelles. Encore mieux, vous pouvez également l'utiliser en tant que générateur de site statique! Nous vous recommandons fortement de l'essayer.

### Quasar {#quasar}

> [Quasar](https://quasar.dev) est une solution complète basée sur Vue qui vous permet de cibler du SPA, du SSR, de la PWA, de l'application mobile, de l'application de bureau et de l'extension de navigateur en utilisant une seule base de code. Il gère non seulement la mise en place du build, mais fournit également une collection complète de composants d'interface utilisateur conformes à Material Design.

Reviewer 2: ```suggestion
[Quasar](https://quasar.dev) est une solution complète basée sur Vue qui vous permet de cibler du SPA, du SSR, de la PWA, de l'application mobile, de l'application de bureau et de l'extension de navigateur en utilisant une seule base de code. Elle gère non seulement la mise en place du build, mais fournit également une collection complète de composants d'interface utilisateur conformes à Material Design.
```

### Vite SSR {#vite-ssr}

> Vite fournit [un support intégré pour le rendu côté serveur de Vue](https://vitejs.dev/guide/ssr.html), mais il est intentionnellement de niveau bas. Si vous souhaitez aller directement avec Vite, consultez [vite-plugin-ssr](https://vite-plugin-ssr.com/), un plugin communautaire qui fait abstraction de nombreux détails difficiles pour vous.

Reviewer 2: ```suggestion
Vite fournit [un support intégré pour le rendu côté serveur de Vue](https://vitejs.dev/guide/ssr.html), mais il est intentionnellement de bas niveau. Si vous souhaitez utiliser directement Vite, consultez [vite-plugin-ssr](https://vite-plugin-ssr.com/), un plugin communautaire qui fait abstraction de nombreux détails complexes pour vous.
```

> Vous pouvez également trouver un exemple de projet Vue + Vite SSR en utilisant une configuration manuelle ici, qui peut servir de base pour la construction. Notez que ceci n'est recommandé que si vous avez de l'expérience avec SSR / outils de build et que vous souhaitez avoir un contrôle complet sur l'architecture de haut niveau.

Reviewer 2: ```suggestion
Vous pouvez également trouver un exemple de projet Vue + Vite SSR en utilisant une configuration manuelle ici, qui peut servir de base pour le build. Notez que ceci n'est recommandé que si vous avez de l'expérience avec SSR / outils de build et que vous souhaitez avoir un contrôle complet sur l'architecture de haut niveau.
```

## Écrire un code convivial pour du SSR {#writing-ssr-friendly-code}

> Indépendamment de votre configuration de construction ou de votre choix de framework de haut niveau, il existe des principes qui s'appliquent à toutes les applications Vue SSR.

Reviewer 2: ```suggestion
Indépendamment de votre configuration de build ou de votre choix de framework de haut niveau, il existe des principes qui s'appliquent à toutes les applications Vue SSR.
```

### Réactivité sur le serveur {#reactivity-on-the-server}

> Pendant le SSR, chaque URL de demande correspond à un état souhaité de notre application. Il n'y a pas d'interaction utilisateur et aucune mise à jour DOM, donc la réactivité n'est pas nécessaire sur le serveur. Par défaut, la réactivité est désactivée pendant le SSR pour une meilleure performance.

Reviewer 2: ```suggestion
Pendant le SSR, chaque URL requêtée correspond à un état souhaité de notre application. Il n'y a pas d'interaction utilisateur et aucune mise à jour du DOM, donc la réactivité n'est pas nécessaire sur le serveur. Par défaut, la réactivité est désactivée pendant le SSR pour une meilleure performance.
```

### Les hooks du cycle de vie du composant {#component-lifecycle-hooks}

> Depuis qu'il n'y a plus de mises à jour dynamiques, les hooks de cycle de vie tels que <span class="options-api">`mounted`</span><span class="composition-api">`onMounted`</span> ou <span class="options-api">`updated`</span><span class="composition-api">`onUpdated`</span> ne seront **PAS** appelées pendant le SSR et ne seront exécutées que côté client.<span class="options-api"> Les seuls hooks qui sont appelés pendant le SSR sont `beforeCreate` et `created`</span>

Reviewer 2: ```suggestion
Depuis qu'il n'y a plus de mises à jour dynamiques, les hooks de cycle de vie tels que <span class="options-api">`mounted`</span><span class="composition-api">`onMounted`</span> ou <span class="options-api">`updated`</span><span class="composition-api">`onUpdated`</span> ne seront **PAS** appelés pendant le SSR et ne seront exécutés que côté client.<span class="options-api"> Les seuls hooks qui sont appelés pendant le SSR sont `beforeCreate` et `created`</span>
```

> Vous devriez éviter le code qui produit des effets secondaires qui nécessitent un nettoyage dans <span class="options-api">`beforeCreate` et `created`</span><span class="composition-api">`setup()` ou dans le scope de `<script setup>`</span>. Un exemple d'effets secondaires est la configuration de minuteries avec `setInterval`. Dans le code exécuté côté client, nous pouvons configurer une minuterie et la démonter dans <span class="options-api">`beforeUnmount`</span><span class="composition-api">`onBeforeUnmount`</span> ou <span class="options-api">`unmounted`</span><span class="composition-api">`onUnmounted`</span>. Cependant, étant donné que les hooks de démontage ne seront jamais appelés pendant le SSR, les minuteries resteront à jamais. Pour éviter cela, déplacez votre code d'effets secondaires dans <span class="options-api">`mounted`</span><span class="composition-api">`onMounted`</span>.

Reviewer 2: ```suggestion
Vous devriez éviter le code qui produit des effets secondaires qui nécessitent un nettoyage dans <span class="options-api">`beforeCreate` et `created`</span><span class="composition-api">`setup()` ou dans la portée de `<script setup>`</span>. Un exemple d'effets secondaires est la configuration de minuteries avec `setInterval`. Dans le code exécuté côté client, nous pouvons configurer une minuterie et la démonter dans <span class="options-api">`beforeUnmount`</span><span class="composition-api">`onBeforeUnmount`</span> ou <span class="options-api">`unmounted`</span><span class="composition-api">`onUnmounted`</span>. Cependant, étant donné que les hooks de démontage ne seront jamais appelés pendant le SSR, les minuteries resteront à jamais. Pour éviter cela, déplacez votre code d'effets secondaires dans <span class="options-api">`mounted`</span><span class="composition-api">`onMounted`</span>.
```

Author: @tlagoda Dans le glossaire, j'ai cru voir qu'on gardait scope ?

Reviewer 2: Bizarrerie 🤣 c'est bien scope dans le glossaire mais c'est traduit portée un peu partout!

> ###  Accès aux API spécifiques à la plateforme {#access-to-platform-specific-apis}

Reviewer 2: ```suggestion
### Accès aux API spécifiques à la plateforme {#access-to-platform-specific-apis}
```

> Universal code cannot assume access to platform-specific APIs, so if your code directly uses browser-only globals like `window` or `document`, they will throw errors when executed in Node.js, and vice-versa.

Reviewer 2: Tu as laissé les blocs non traduits ici

Le code universel ne peut pas supposer l'accès aux API spécifiques à la plateforme, donc si votre code utilise directement des éléments spécifiques au navigateur comme `window` ou `document`, ils lanceront des erreurs lorsqu'ils seront exécutés dans Node.js et vice-versa.

For tasks that are shared between server and client but with different platform APIs, it's recommended to wrap the platform-specific implementations inside a universal API, or use libraries that do this for you. For example, you can use [`node-fetch`](https://github.com/node-fetch/node-fetch) to use the same fetch API on both server and client.

> Pour les tâches partagées entre le serveur et le client mais avec des API de plateforme différentes, il est recommandé de wraper les implémentations spécifiques à la plateforme dans une API universelle ou d'utiliser des bibliothèques qui le font pour vous. Par exemple, vous pouvez utiliser [`node-fetch`](https://github.com/node-fetch/node-fetch) pour utiliser la même API fetch sur le serveur et le client.

Reviewer 2: ```suggestion
Pour les tâches partagées entre le serveur et le client mais avec des API de plateforme différentes, il est recommandé d'envelopper les implémentations spécifiques à la plateforme dans une API universelle ou d'utiliser des bibliothèques qui le font pour vous. Par exemple, vous pouvez utiliser [`node-fetch`](https://github.com/node-fetch/node-fetch) pour utiliser la même API fetch sur le serveur et le client.
```

Pour les API spécifiques au navigateur, la démarche courante consiste à y accéder de manière paresseuse dans des hooks de cycle de vie client uniquement tels que <span class="options-api">`mounted`</span><span class="composition-api">`onMounted`</span>.

> Notez que si une bibliothèque tiers n'est pas écrite en vue d'une utilisation universelle, il peut être difficile de l'intégrer dans une application rendue côté serveur. Vous _pourriez_ être en mesure de la faire fonctionner en simulant certaines des globales, mais ce serait un bricolage et peut interférer avec le code de détection de l'environnement d'autres bibliothèques.

Reviewer 2: ```suggestion
Notez que si une bibliothèque tiers n'est pas écrite en vue d'une utilisation universelle, il peut être difficile de l'intégrer dans une application rendue côté serveur. Vous _pourriez_ être en mesure de la faire fonctionner en simulant certaines des globales, mais ce serait un bricolage et pourrait interférer avec le code de détection de l'environnement d'autres bibliothèques.
```

> ### Pollution d'État à demande croisée {#cross-request-state-pollution}

Reviewer 2: ```suggestion
### Pollution d'état par demandes croisées {#cross-request-state-pollution}
```

@edimitchel une idée ? 🤣

Reviewer: ta proposition me convient @tlagoda !

Dans le chapitre de Gestion d'État, nous avons introduit un [patron de gestion d'état simple en utilisant les API de réactivité](state-management.html#simple-state-management-with-reactivity-api). Dans un contexte SSR, ce modèle nécessite des ajustements supplémentaires.

> Le modèle déclare un état partagé dans le scope d'un module JavaScript. Cela les rend **singletons** - c'est-à-dire qu'il n'y a qu'une seule instance de l'objet réactif tout au long de la vie de notre application. Cela fonctionne comme prévu dans une application Vue côté client, car les modules de notre application sont initialisés à nouveau pour chaque visite de page navigateur.

Reviewer 2: ```suggestion
Le modèle déclare un état partagé dans la portée d'un module JavaScript. Cela les rend **singletons** - c'est-à-dire qu'il n'y a qu'une seule instance de l'objet réactif tout au long de la vie de notre application. Cela fonctionne comme prévu dans une application Vue côté client, car les modules de notre application sont initialisés à nouveau pour chaque visite de page navigateur.
```

Author: @tlagoda Dans le glossaire, j'ai cru voir qu'on gardait scope ?

Reviewer: Effectivement, mais en vrai, la question de se pose, portée me semble assez limpide comme traduction de scope.
On peut peut-être remettre en question ce choix.

Author: Je la trouve bien cette traduction, allons y sur ça

> Cependant, dans un contexte SSR, les modules d'application sont généralement initialisés une seule fois sur le serveur, lorsque le serveur démarre. Les mêmes instances de modules seront réutilisées à travers plusieurs demandes de serveur, et donc nos objets d'état singletons. Si nous modifions l'état partagé singleton avec des données spécifiques à un utilisateur, cela peut être accidentellement fuite à une demande provenant d'un autre utilisateur. Nous appelons cela une **pollution d'état cross-request.**

Reviewer 2: ```suggestion
Cependant, dans un contexte SSR, les modules d'application sont généralement initialisés une seule fois sur le serveur, lorsque le serveur démarre. Les mêmes instances de modules seront réutilisées à travers plusieurs demandes de serveur, et donc nos objets d'état singletons. Si nous modifions l'état partagé singleton avec des données spécifiques à un utilisateur, cela peut être accidentellement fuir vers une demande provenant d'un autre utilisateur. Nous appelons cela une **pollution d'état cross-request.**
```

Techniquement, nous pouvons ré-initialiser tous les modules JavaScript à chaque demande, tout comme nous le faisons dans les navigateurs. Cependant, l'initialisation des modules JavaScript peut être coûteuse, ce qui affecterait significativement les performances du serveur.

The recommended solution is to create a new instance of the entire application - including the router and global stores - on each request. Then, instead of directly importing it in our components, we provide the shared state using [app-level provide](/guide/components/provide-inject.html#app-level-provide) and inject it in components that need it:

> La solution recommandée est de créer une nouvelle instance de l'application entière - y compris le routeur et les stores globaux - à chaque demande. Ensuite, au lieu de l'importer directement dans nos composants, nous fournissons l'état partagé en utilisant [le fournisseur de niveau d'application](/guide/components/provide-inject.html#app-level-provide) et l'injecterons dans les composants qui en ont besoin :

Reviewer 2: Pas sûr pour la trad du titre du lien, est-ce qu'on traduit `provide` ?

Author: @tlagoda Je viens de voir sur la page Provide / Inject, c'est traduit : Provide au niveau de l'application ! j'avais pas vu, je fixe

```js
// app.js (partagé entre le serveur et le client)
import { createSSRApp } from 'vue'
import { createStore } from './store.js'

// appelé sur chaque demande
export function createApp() {
  const app = createSSRApp(/* ... */)

>   // créer une nouvelle instance de store par requête

Reviewer 2: ```suggestion
  // crée une nouvelle instance de store par requête
```

  const store = createStore(/* ... */)

>   // fournir un store au niveau de l'application

Reviewer 2: ```suggestion
  // fournit un store au niveau de l'application
```

  app.provide('store', store)
  // expose également le store à des fins d'hydratation
  return { app, store }
}
```

Les bibliothèques de gestion d'état comme Pinia sont conçues dans cet esprit. Consultez le [guide SSR de Pinia](https://pinia.vuejs.org/ssr/) pour plus de détails.

### Déséquilibre de l'hydratation {#hydration-mismatch}

Si la structure du DOM du HTML pré-rendu ne correspond pas à la sortie attendue de l'application côté client, il y aura une erreur de désaccord d'hydratation. Le désaccord d'hydratation est généralement introduit par les causes suivantes :

> 1. Le modèle contient une structure de nesting HTML non valide, et le HTML rendu a été "corrigé" par le comportement de parsing HTML natif du navigateur. Par exemple, une erreur courante est que [`<div>` ne peut pas être placé à l'intérieur de `<p>`](https://stackoverflow.com/questions/8397852/why-cant-the-p-tag-contain-a-div-tag-inside-it):

Reviewer 2: ```suggestion
1. Le modèle contient une structure HTML imbriqué non valide, et le HTML rendu a été "corrigé" par le comportement d'analyse syntaxique du HTML natif du navigateur. Par exemple, une erreur courante est que [`<div>` ne peut pas être placée à l'intérieur de `<p>`](https://stackoverflow.com/questions/8397852/why-cant-the-p-tag-contain-a-div-tag-inside-it):
```

   ```html
   <p><div>hi</div></p>
   ```

>    Si nous produisons ceci dans notre HTML rendu par le serveur, le navigateur terminera le premier `<p>` lorsque `<div>` sera rencontré et le parser en la structure de DOM suivante :

Reviewer 2: ```suggestion
   Si nous produisons ceci dans notre HTML rendu par le serveur, le navigateur terminera le premier `<p>` lorsque `<div>` sera rencontré et l'analysera en la structure de DOM suivante :
```

   ```html
   <p></p>
   <div>hi</div>
   <p></p>
   ```

2. Les données utilisées lors de la mise en forme contiennent des valeurs générées aléatoirement. Étant donné que la même application s'exécutera deux fois - une fois sur le serveur et une fois sur le client - les valeurs aléatoires ne sont pas garanties d'être les mêmes entre les deux exécutions. Il existe deux façons d'éviter les incohérences causées par les valeurs aléatoires :

   1. Utilisez `v-if` + `onMounted` pour rendre uniquement sur le client la partie dépendante de valeurs aléatoires. Votre framework peut également disposer de fonctionnalités intégrées pour faciliter cela, par exemple le composant `<ClientOnly>` de VitePress.

   2. Use a random number generator library that supports generating with seeds, and guarantee the server run and the client run are using the same seed (e.g. by including the seed in serialized state and retrieving it on the client).

>    Utilisez une bibliothèque de générateur de nombres aléatoires qui prend en charge la génération avec des seeds, et garantissez que l'exécution du serveur et l'exécution du client utilisent la même seed (par exemple en incluant la seed dans l'état sérialisé et en la récupérant sur le client).

Reviewer 2: ```suggestion
   Utilisez une bibliothèque de générateur de nombres aléatoires qui prend en charge la génération avec des seeds, et assurez-vous que l'exécution du serveur et l'exécution du client utilisent la même seed (par exemple en incluant la seed dans l'état sérialisé et en la récupérant sur le client).
```

> 3. Le serveur et le client se trouvent dans des fuseaux horaires différents. Parfois, nous pouvons vouloir convertir un horodatage en temps local de l'utilisateur. Cependant, le fuseau horaire lors de l'exécution du serveur et le fuseau horaire lors de l'exécution du client ne sont pas toujours les mêmes, et nous ne pouvons pas toujours connaître de manière fiable le fuseau horaire de l'utilisateur lors de l'exécution du serveur. Dans de tels cas, la conversion de l'heure locale doit également être effectuée en tant qu'opération uniquement côté client.

Reviewer 2: ```suggestion
3. Le serveur et le client se trouvent dans des fuseaux horaires différents. Parfois, nous pouvons vouloir convertir un timestamp en temps local de l'utilisateur. Cependant, le fuseau horaire lors de l'exécution du serveur et le fuseau horaire lors de l'exécution du client ne sont pas toujours les mêmes, et nous ne pouvons pas toujours connaître de manière fiable le fuseau horaire de l'utilisateur lors de l'exécution du serveur. Dans de tels cas, la conversion de l'heure locale doit également être effectuée en tant qu'opération uniquement côté client.
```

```

You need to inject the teleport markup into the correct location in your final page HTML similar to how you need to inject the main app markup.

> Vous devez injecter le markup de téléportation dans l'emplacement correct dans votre HTML de page finale, similaire à comment vous devez injecter le markup principal de l'application.

Reviewer 2: ```suggestion
Vous devez injecter le markup de téléportation dans l'emplacement correct dans votre HTML de page finale, tout comme vous devez injecter le markup principal de l'application.
```

:::tip

> Évitez de cibler `body` lorsque vous utilisez Teleports et SSR ensemble - généralement, `<body>` contiendra d'autres contenus rendus sur le serveur ce qui rend impossible pour Teleports de déterminer le bon emplacement de départ pour l'hydratation.

Reviewer 2: ```suggestion
Évitez de cibler `body` lorsque vous utilisez Teleports et le SSR ensemble - généralement, `<body>` contiendra d'autres contenus rendus sur le serveur ce qui rend impossible pour Teleports de déterminer le bon emplacement de départ pour l'hydratation.
```
